x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/UpdateIndexMigrationVersionAction.java

```diff
@@ -0,0 +1,191 @@
+            return indexMigrationVersion;
+        }
+
+        public String getIndexName() {
+            return indexName;
+        }
+    }
+
+    public static class TransportAction extends TransportMasterNodeAction<Request, UpdateIndexMigrationVersionResponse> {
+        private final MasterServiceTaskQueue<UpdateIndexMigrationVersionTask> updateIndexMigrationVersionTaskQueue;
+
+        @Inject
+        public TransportAction(
+            TransportService transportService,
+            ClusterService clusterService,
+            ThreadPool threadPool,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver
+        ) {
+            super(
+                UpdateIndexMigrationVersionAction.NAME,
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                Request::new,
+                indexNameExpressionResolver,
+                UpdateIndexMigrationVersionResponse::new,
+                threadPool.executor(ThreadPool.Names.MANAGEMENT)
+            );
+            this.updateIndexMigrationVersionTaskQueue = clusterService.createTaskQueue(
+                "update-index-migration-version-task-queue",
+                Priority.HIGH,
+                UPDATE_INDEX_MIGRATION_VERSION_TASK_EXECUTOR
+            );
+        }
+
+        private static final SimpleBatchedExecutor<UpdateIndexMigrationVersionTask, Void> UPDATE_INDEX_MIGRATION_VERSION_TASK_EXECUTOR =
+            new SimpleBatchedExecutor<>() {
+                @Override
+                public Tuple<ClusterState, Void> executeTask(UpdateIndexMigrationVersionTask task, ClusterState clusterState) {
+                    return Tuple.tuple(task.execute(clusterState), null);
+                }
+
+                @Override
+                public void taskSucceeded(UpdateIndexMigrationVersionTask task, Void unused) {
+                    task.listener.onResponse(null);
+                }
+            };
+
+        static class UpdateIndexMigrationVersionTask implements ClusterStateTaskListener {
+            private final ActionListener<Void> listener;
+            private final int indexMigrationVersion;
+            private final String indexName;
+
+            UpdateIndexMigrationVersionTask(ActionListener<Void> listener, int indexMigrationVersion, String indexName) {
+                this.listener = listener;
+                this.indexMigrationVersion = indexMigrationVersion;
+                this.indexName = indexName;
+            }
+
+            ClusterState execute(ClusterState currentState) {
+                IndexMetadata.Builder indexMetadataBuilder = IndexMetadata.builder(currentState.metadata().getIndices().get(indexName));
+                indexMetadataBuilder.putCustom(
+                    MIGRATION_VERSION_CUSTOM_KEY,
+                    Map.of(MIGRATION_VERSION_CUSTOM_DATA_KEY, Integer.toString(indexMigrationVersion))
```

Comment:
nit: This is just a suggestion, I'm not firm about it:

I think it can simplify future possible troubleshooting, if we make this action to "increase the migration version by 1 iff the current migration version is X, otherwise fail" instead of "set migration version to Y". WDYT?

Reply:
Your analysis looks right to me. There shouldn't be concurrent actions to update the index metadata because of the persistent task framework guarantees.
I'm fine to live it like this.